### ISSUE TITLE
OpenSUSE test failures on Python 3.7

### ISSUE THREAD

**Issue by jayvdb** [TriOptima/tri.struct#4]
Some of the openSUSE builds are failing with
```
[   13s] 
[   13s] =================================== FAILURES ===================================
[   13s] ______________________________ test_copy[CStruct] ______________________________
[   13s] 
[   13s] Struct = <class 'tri_struct.Struct'>
[   13s] 
[   13s]     def test_copy(Struct):
[   13s]         s = Struct(a=1, b=2, c=3)
[   13s]         q = s.copy()
[   13s]     
[   13s]         assert 'a' in q
[   13s]         assert 'b' in q
[   13s]         assert 'c' in q
[   13s]     
[   13s]         q.x = 6
[   13s]         with pytest.raises(AttributeError) as e:
[   13s]             # noinspection PyStatementEffect
[   13s]             s.x
[   13s]     
[   13s] >       assert "'Struct' object has no attribute 'x'" in str(e)
[   13s] E       assert "'Struct' object has no attribute 'x'" in '<ExceptionInfo AttributeError tblen=1>'
[   13s] E        +  where '<ExceptionInfo AttributeError tblen=1>' = str(<ExceptionInfo AttributeError tblen=1>)
[   13s] 
[   13s] tests/test_struct.py:84: AssertionError
[   13s] _______________________ test_no_longer_has_dict[CStruct] _______________________
[   13s] 
[   13s] Struct = <class 'tri_struct.Struct'>
[   13s] 
[   13s]     def test_no_longer_has_dict(Struct):
[   13s]         s = Struct()
[   13s]         with pytest.raises(AttributeError) as e:
[   13s]             s.__dict__
[   13s] >       assert "'Struct' object has no attribute '__dict__'" in str(e)
[   13s] E       assert "'Struct' object has no attribute '__dict__'" in '<ExceptionInfo AttributeError tblen=1>'
[   13s] E        +  where '<ExceptionInfo AttributeError tblen=1>' = str(<ExceptionInfo AttributeError tblen=1>)
[   13s] 
[   13s] tests/test_struct.py:109: AssertionError
[   13s] ______________________________ test_hash[CStruct] ______________________________
[   13s] 
[   13s] Struct = <class 'tri_struct.Struct'>
[   13s] 
[   13s]     def test_hash(Struct):
[   13s]         s = Struct(x=17)
[   13s]         with pytest.raises(TypeError) as e:
[   13s]             hash(s)
[   13s]         if platform.python_implementation() == "PyPy":
[   13s]             assert "" in str(e)
[   13s]         else:
[   13s] >           assert "unhashable type: 'Struct'" in str(e)
[   13s] E           assert "unhashable type: 'Struct'" in '<ExceptionInfo TypeError tblen=1>'
[   13s] E            +  where '<ExceptionInfo TypeError tblen=1>' = str(<ExceptionInfo TypeError tblen=1>)
[   13s] 
[   13s] tests/test_struct.py:141: AssertionError
[   13s] _____________________________ test_copy[PyStruct] ______________________________
[   13s] 
[   13s] Struct = <class 'tri_struct.Struct'>
[   13s] 
[   13s]     def test_copy(Struct):
[   13s]         s = Struct(a=1, b=2, c=3)
[   13s]         q = s.copy()
[   13s]     
[   13s]         assert 'a' in q
[   13s]         assert 'b' in q
[   13s]         assert 'c' in q
[   13s]     
[   13s]         q.x = 6
[   13s]         with pytest.raises(AttributeError) as e:
[   13s]             # noinspection PyStatementEffect
[   13s]             s.x
[   13s]     
[   13s] >       assert "'Struct' object has no attribute 'x'" in str(e)
[   13s] E       assert "'Struct' object has no attribute 'x'" in '<ExceptionInfo AttributeError tblen=3>'
[   13s] E        +  where '<ExceptionInfo AttributeError tblen=3>' = str(<ExceptionInfo AttributeError tblen=3>)
[   13s] 
[   13s] tests/test_struct.py:84: AssertionError
[   13s] ______________________ test_no_longer_has_dict[PyStruct] _______________________
[   13s] 
[   13s] Struct = <class 'tri_struct.Struct'>
[   13s] 
[   13s]     def test_no_longer_has_dict(Struct):
[   13s]         s = Struct()
[   13s]         with pytest.raises(AttributeError) as e:
[   13s]             s.__dict__
[   13s] >       assert "'Struct' object has no attribute '__dict__'" in str(e)
[   13s] E       assert "'Struct' object has no attribute '__dict__'" in '<ExceptionInfo AttributeError tblen=3>'
[   13s] E        +  where '<ExceptionInfo AttributeError tblen=3>' = str(<ExceptionInfo AttributeError tblen=3>)
[   13s] 
[   13s] tests/test_struct.py:109: AssertionError
[   13s] _____________________________ test_hash[PyStruct] ______________________________
[   13s] 
[   13s] Struct = <class 'tri_struct.Struct'>
[   13s] 
[   13s]     def test_hash(Struct):
[   13s]         s = Struct(x=17)
[   13s]         with pytest.raises(TypeError) as e:
[   13s]             hash(s)
[   13s]         if platform.python_implementation() == "PyPy":
[   13s]             assert "" in str(e)
[   13s]         else:
[   13s] >           assert "unhashable type: 'Struct'" in str(e)
[   13s] E           assert "unhashable type: 'Struct'" in '<ExceptionInfo TypeError tblen=1>'
[   13s] E            +  where '<ExceptionInfo TypeError tblen=1>' = str(<ExceptionInfo TypeError tblen=1>)
[   13s] 
[   13s] tests/test_struct.py:141: AssertionError
[   13s] - generated xml file: /home/abuild/rpmbuild/BUILD/tri.struct-3.0.1/testreport.xml -
```

https://build.opensuse.org/package/show/home:jayvdb:py-new/python-tri.struct

It looks like the Python 3.6 jobs are passing, and it is the Python 3.7 jobs failing.

**Comment by boxed** [comment]
Hmm.. what version of pytest do you have? I see that we don't lock the version of pytest in test_requirements.txt and I think I remember them changing how caught exception objects are handled.

**Comment by jayvdb** [comment]
AH, yes that is possibly the problem.  It is using pytest 5.

If I instead install pytest 5, but using `python3 setup.py test` , no tests are run

```
/usr/bin/python3 setup.py test
[   15s] running test
[   15s] running egg_info
[   15s] writing lib/tri.struct.egg-info/PKG-INFO
[   15s] writing dependency_links to lib/tri.struct.egg-info/dependency_links.txt
[   15s] writing top-level names to lib/tri.struct.egg-info/top_level.txt
[   15s] reading manifest file 'lib/tri.struct.egg-info/SOURCES.txt'
[   15s] reading manifest template 'MANIFEST.in'
[   15s] writing manifest file 'lib/tri.struct.egg-info/SOURCES.txt'
[   15s] running build_ext
[   15s] building 'tri_struct._cstruct' extension
[   15s] gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -DOPENSSL_LOAD_CONF -fwrapv -O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -fPIC -I/usr/include/python3.7m -c lib/tri_struct/_cstruct.c -o build/temp.linux-s390x-3.7/lib/tri_struct/_cstruct.o
[   15s] gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -DOPENSSL_LOAD_CONF -fwrapv -O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -fPIC -I/usr/include/python3.7m -c lib/tri_struct/_typespec.c -o build/temp.linux-s390x-3.7/lib/tri_struct/_typespec.o
[   15s] gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -DOPENSSL_LOAD_CONF -fwrapv -O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -O2 -g -m64 -fmessage-length=0 -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -fPIC -I/usr/include/python3.7m -c lib/tri_struct/_utils.c -o build/temp.linux-s390x-3.7/lib/tri_struct/_utils.o
[   15s] gcc -pthread -shared build/temp.linux-s390x-3.7/lib/tri_struct/_cstruct.o build/temp.linux-s390x-3.7/lib/tri_struct/_typespec.o build/temp.linux-s390x-3.7/lib/tri_struct/_utils.o -L/usr/lib64 -lpython3.7m -o /home/abuild/rpmbuild/BUILD/tri.struct-3.0.1/lib/tri_struct/_cstruct.cpython-37m-s390x-linux-gnu.so
[   15s] 
[   15s] ----------------------------------------------------------------------
[   15s] Ran 0 tests in 0.000s
```

**Comment by boxed** [comment]
Fixed in 63075abd5e10cebb405c2060c26ec4ad537b263b

**Comment by jayvdb** [comment]
Thanks.  Packaged at https://build.opensuse.org/project/show/home:jayvdb:py-tri , and I'll get the other tests working too (some disabled atm as Django needs a bit of fiddling to get tests working).  Main goal is actually getting mutmut onto openSUSE.
